soroban-sdk/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,294 @@
+/// let one = G1Affine::from_bytes(bytesn!(&env, 0x17f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb08b3f481e3aaa0f1a09e30ed741d8ae4fcf5e095d5d00af600db18cb2c04b3edd03cc744a2888ae40caa232946c5e7e1));
+/// let res = bls12_381.g1_add(&zero, &one);
+/// assert_eq!(res, one);
+/// ```
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct G1Affine(BytesN<96>);
+
+/// # `G2Affine` is a point in the G2 group (subgroup defined over the quadratic
+/// extension field `Fq2`) of the BLS12-381 elliptic curve
+///
+/// # Serialization:
+/// - The 192 bytes represent the **uncompressed encoding** of a point in G2.
+///   The bytes consist of `be_bytes(X_c1) || be_bytes(X_c0) || be_bytes(Y_c1)
+///   || be_bytes(Y_c0)` (`||` is concatenation), where 'X' and 'Y' are the two
+///   coordinates, each being an extension field element `Fp2` and `c0`, `c1`
+///   are components of `Fp2` (each being `Fp`).
+/// - The most significant three bits (bits 0-3) of the first byte are reserved
+///   for encoding flags:
+///   - compression_flag (bit 0): Must always be set (1), as only uncompressed
+///     points are supported.
+///   - infinity_flag (bit 1): Set if the point is the point at infinity (zero
+///     point), in which case all other bits must be zero.
+///   - sort_flag (bit 2): Must always be unset (0).
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct G2Affine(BytesN<192>);
+
+/// # `Fp` represents an element of the base field `Fq` of the BLS12-381 elliptic
+/// curve
+///
+/// # Serialization:
+/// - The 48 bytes represent the **big-endian encoding** of an element in the
+///   field `Fp`. The value is serialized as a big-endian integer.
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Fp(BytesN<48>);
+
+/// # `Fp2` represents an element of the quadratic extension field `Fq2` of the
+/// BLS12-381 elliptic curve
+///
+/// # Serialization:
+/// - The 96 bytes represent the **big-endian encoding** of an element in the
+///   field `Fp2`. The bytes consist of `be_bytes(c1) || be_bytes(c0)` (`||` is
+///   concatenation), where `c0` and `c1` are the two `Fp` elements (the real
+///   and imaginary components).
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Fp2(BytesN<96>);
+
+impl_bytesn_repr!(G1Affine, 96);
+impl_bytesn_repr!(G2Affine, 192);
+impl_bytesn_repr!(Fp, 48);
+impl_bytesn_repr!(Fp2, 96);
+
+impl Bls12_381 {
+    pub(crate) fn new(env: &Env) -> Bls12_381 {
+        Bls12_381 { env: env.clone() }
+    }
+
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+
+    // g1
+
+    /// Checks if a point `p` in G1 is in the correct subgroup.
+    pub fn g1_is_in_subgroup(&self, p: &G1Affine) -> bool {
+        let env = self.env();
+        let res = internal::Env::bls12_381_check_g1_is_in_subgroup(env, p.to_object())
+            .unwrap_infallible();
+        res.into()
+    }
+
+    /// Adds two points `p0` and `p1` in G1.
+    pub fn g1_add(&self, p0: &G1Affine, p1: &G1Affine) -> G1Affine {
```

Comment:
It looks like most of the functions on the Bls12_381 type orient around the bls types. For example, this one here `g1_add` could be a function on g1.

In general I quite like these style of APIs where modules contain the functionality and types are just data.

However it introduces some awkwardness when writing contracts.

If a developer writes code that uses the G1, or other, types, then they might be passing them around. If all the operations are on the env, they need to get the env either from the G1 or by passing it around too. In rust docs it also becomes harder to discover all the operations that can be done on a g1, g2, etc.

For some the operations, like add, we can impl the core lib's traits, e.g. Add.

```rust
fn do_things(a: G1Affine, b: G1Affine) /* -> ... */ {
    let c = a.env().crypto().bls12_381().g1_add(a, b);
    // ...
}
```
vs
```rust
fn do_things(a: G1Affine, b: G1Affine) /* -> ... */ {
    let c = a.add(b);
    // ...
}
```

Reply:
I might be missing something... but how would you give the Add trait access to Env? We still need Env to add the two points right?

Reply:
The Add trait doesn't need to be given the Env. The Env already lives inside G1Affine, so ti has access to it.

Reply:
For example:

```rust
impl Bls12_381 {
    pub fn g1_add(&self, p0: &G1Affine, p1: &G1Affine) -> G1Affine {
        let env = self.env();
        let bin = internal::Env::bls12_381_g1_add(env, p0.to_object(), p1.to_object())
            .unwrap_infallible();
        unsafe { G1Affine::from_bytes(<BytesN<96>>::unchecked_new(env.clone(), bin)) }
    }
```

Becomes:

```rust
impl G1Affine {
    pub fn env(&self) -> &Env {
        self.0.env()
    }
}

impl core::ops::Add for G1Affine {
    type Output = G1Affine;

    fn add(self, rhs: Self) -> Self::Output {
        let env = self.env();
        let bin = internal::Env::bls12_381_g1_add(env, self.to_object(), rhs.to_object())
            .unwrap_infallible();
        unsafe { G1Affine::from_bytes(<BytesN<96>>::unchecked_new(env.clone(), bin)) }
    }
}
```

Reply:
Ah right. My IDE was taking me into the arkworks library for G1Affine and I got confused.

So you're advocating for doing this and removing `g1_add` instead of having both options right?

Reply:
🤔 I was advocating for at least having the fns on the types, but I think it would also work to have both, the `G1Affine::add` can call `env.g1_add(...)` if you preferred to do that.

Reply:
Also don't forget to implement `CheckedAdd` if we're going this route. Though the downside is that the documentation is not as discoverable, so maybe instead of a trait we should have a separate member function.

Reply:
We can provide both if we're concerned about discoverability. e.g.:

```rust
impl core::ops::Add for G1Affine {
    type Output = Self;

    fn add(self, rhs: Self) -> Self::Output {
        self.add(rhs)
    }
}

impl G1Affine {
    pub fn add(self, rhs: Self) -> Self {
        self.env().crypto().bls12_381().g1_add(&self, &rhs)
    }

    pub fn checked_add(self, rhs: Self) -> Option<Self> {
        self.env().crypto().bls12_381().g1_checked_add(&self, &rhs)
    }
}
```

There's no `CheckedAdd` trait, so it'll only be a fn anyway.

Reply:
Looks like the functions here would lend to implementing these `core::ops` traits:
- `Add`
- `AddAssign`
- `Mul`
- `MulAssign`
- `Sub`
- `SubAssign`

Reply:
Ah I didn't realize rhs could be templated. Updated

Reply:
Reminder there's also the `Fr` type.

Reply:
Did you mean the `Fp` type? I'll add the map_to_* functions for those, but not sure if that's what you meant.

Reply:
Yes that too, I was meaning the fr_add functions, but I see now there's no Fr type. Why is that?

Reply:
Hmm yeah we should probably add a Fr type that wraps a U256, along with those methods.

Reply:
Although I'm not sure if this improves usability. I need to look into this.

Reply:
+1 I don't think we should push this in a direction where the types are a hinderance and provide no utility.

Reply:
Yeah in this case wrapping the U256 doesn't add much imo because the host uses U256. If we wanted to limit the possible set of operations on those U256s, then wrapping it in a type makes sense, but it doesn't look like any limitation is required.

I would choose to leave as is for now.

Reply:
One more thought to throw in the mix, do the `fr_add`/`sub`/`mul` fns all behave the same way as the `u256_add`/`sub`/`mul`? I assume no, otherwise we wouldn't have them. It could be a footgun for someone to see the G1Affine type, pass it around, call add on it, then see the U256 type, pass it around and call add on it.

Reply:
For comparison the Rust bls lib seems to use a type named `Scalar` for the `Fr` associated type: https://docs.rs/bls12_381/latest/bls12_381/struct.Bls12.html

Reply:
I'm going to merge this PR to avoid bitrot but will open a follow up issue to implement before we release.

Reply:
https://github.com/stellar/rs-soroban-sdk/issues/1352

Reply:
This is a good catch, thank you @leighmcculloch and @sisuresh for filing the issue.
`Fr` is the 255-bit scalar with a fixed modulus (equal to the subgroup order `r`) and all its operations are modulo, it's just convenient to share the representation with U256. I will work on the followup fix.